File: MinStack.java
import java.util.ArrayList;
public class MinStack {
    public ArrayList<Integer> stack = new ArrayList<Integer>();
    public MinStack() {}
    
    public void push(int val) {
        this.stack.add(val);
    }
    
    public void pop() {
        stack.remove(stack.size()-1);
    }
    
    public int top() {
        return stack.get(stack.size()-1);
    }
    
    public int getMin() {
        int smallest = Integer.MAX_VALUE;
        for(int i = 0; i < stack.size(); i++){
            if(stack.get(i) < smallest)
                smallest = stack.get(i);
        }

        return smallest;
    }

    public static void main(String[] args) {
        MinStack newStack = new MinStack();
        newStack.push(-2);
        newStack.push(0);
        newStack.push(-3);
        System.out.println(newStack.getMin());// return -3
        newStack.pop();
        System.out.println(newStack.top());   // return 0
        System.out.println(newStack.getMin());// return -2
    }
}

//overall O(n) implementation only for the getMin() function. 

File: MinStack.py
import collections
class MinStack:
    def __init__(self):
        self.stack1 = collections.deque
        self.stack2 = collections.deque

    def push(self, val: int) -> None:
        self.stack1.append(val)
        if len(self.stack2) == 0 or val <= self.stack2[len(self.stack2)-1]:
            self.stack2.append(val)
 
    def pop(self) -> None:
        p = self.stack1.pop()
        if p == self.stack2[len(self.stack2)-1]:
            self.stack2.pop()
    def top(self) -> int:
        return self.stack1[len(self.stack1)-1]

    def getMin(self) -> int:
        return self.stack2[len(self.stack2)-1]

#Python implementation, O(1), better than the solution using arraylist